Issue #3793  - Externalize ua-parser regex file

userkci opened the pull request:
This adds an optional configuration parameter for the UaParserEnrichment which can be used to specify an alternate file to provide parsing regexes.  If omitted, the enrichment uses the regexes.yml packaged in the jar, preserving the existing behavior.

The new parameter is `regexFile` and is a URI pointing to the regex file to use in place of the default. There is/will be an accompanying PR for iglu-central for the configuration update.

This PR also includes a change to the mechanism for signaling what files (if any) require local caching by the enrichment process, as the regex file (if configured) is another file that needs caching.  Hopefully, this will prove useful for supporting similar externalization of data for other enrichments.

userkci commented:
As if the title isn't a tipoff, this addresses issue #3793.

snowplow/iglu-central#791 is the corresponding schema update.

snowplowcla commented:
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://github.com/snowplow/snowplow/wiki/CLA to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

BenFradet commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentRegistry.scala:
```diff
@@ -194,6 +196,15 @@ object EnrichmentRegistry {
+   * local path.
+   */
+  val filesToCache: List[(URI, String)] =
+    configs.values.flatMap(_.filesToCache).toList
```
Excellent :+1:

BenFradet commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/UaParserEnrichment.scala:
```diff
@@ -33,34 +37,77 @@ import ua_parser.Client
 object UaParserEnrichmentConfig extends ParseableEnrichment {
+  implicit val formats = DefaultFormats
 
   val supportedSchema = SchemaCriterion("com.snowplowanalytics.snowplow", "ua_parser_config", "jsonschema", 1, 0)
```
this will need to be version 2

BenFradet commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/UaParserEnrichment.scala:
```diff
@@ -33,34 +37,77 @@ import ua_parser.Client
+    customRulefile.map(List(_)).getOrElse(List.empty)
 
-  val uaParser = new Parser()
+  lazy val uaParser: Parser =
```
This should be a validation based on the result of:

```scala
else {
  for {
    input <- Validation.fromTryCatch(new FileInputStream(filesToCache.head._2))
   parser <- try {
     new Parser(input).success
    } catch {
      case NonFatal(e) => e.failure
    } finally {
      input.close()
    }
  } yield parser
}
```

userkci commented:
I've updated the configuration to track the changes to the schema, as referenced in the iglu-central PR.  With the error handling, I've reworked the parser initialization since both Parser() and Parser(InputStream) can throw.  The parser is now wrapped in a validation.  This obviously then had a knock-on effect on the user agent parsing.  I went with reporting the failure on parse, but it unfortunately means all events will fail on this.  Not really content with this, but not sure how to best to handle here.

chuwy commented on 3-enrich/scala-common-enrich/src/test/scala/com.snowplowanalytics.snowplow.enrich.common/enrichments/EnrichmentRegistrySpec.scala:
```diff
@@ -0,0 +1,46 @@
+package com.snowplowanalytics.snowplow.enrich
```
Missing license header.

chuwy commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/enrichments.scala:
```diff
@@ -37,7 +39,20 @@ import utils.ScalazJson4sUtils
+   * second indicates the local path where the enrichment expects
+   * to find the file.
+   */
+  def filesToCache(): List[(URI, String)] = List.empty
```
We usually add trailing `()` only to site-effecting `Unit` methods.

chuwy commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/UaParserEnrichment.scala:
```diff
@@ -106,22 +163,23 @@ case object UaParserEnrichment extends Enrichment {
+        case NonFatal(e) => "Exception parsing useragent [%s]: [%s]".format(useragent, e.getMessage).fail
+      }
+    } yield {
+      // To display useragent version
```
Can we factor out it into a separate function. I feel bad about having such big expression in `yield`.

chuwy commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/UaParserEnrichment.scala:
```diff
@@ -106,22 +163,23 @@ case object UaParserEnrichment extends Enrichment {
+      c <- try {
+        parser.parse(useragent).success
+      } catch {
+        case NonFatal(e) => "Exception parsing useragent [%s]: [%s]".format(useragent, e.getMessage).fail
```
We prefer string interpolation over `String#format`.

```scala
s"Exception parsing useragent [$useragent]: [${e.getMessage}]".failure
```

chuwy commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/UaParserEnrichment.scala:
```diff
@@ -33,34 +37,87 @@ import ua_parser.Client
+    val parser = for {
+      input <- Validation.fromTryCatch(customRulefile.map(f => new FileInputStream(f._2)))
+      p <- try {
+        constructParser(input).success
```
Cannot it also be `fromTryCatch`?

BenFradet replied:
actually I think we shouldn't use `fromTryCatch` at all since it also catches fatal errors in the version of scalaz we use

chuwy replied:
Good point.

chuwy commented on 3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/UaParserEnrichment.scala:
```diff
@@ -33,34 +37,87 @@ import ua_parser.Client
+  private def getUri(uri: String): ValidatedMessage[URI] =
+    ConversionUtils
+      .stringToUri(uri)
+      .flatMap(_ match {
```
```scala
.flatMap {
  case Some(u) => ???
  case None => ???
}
```

userkci commented:
I've made the updates, including the styling.  Because if I'm updating, why not anyway? ;)

BenFradet commented:
Thanks a lot, merged into the release branch for R109 :+1:
